document-store/src/main/java/org/hypertrace/core/documentstore/mongo/query/parser/MongoLetClauseBuilder.java

```diff
@@ -0,0 +1,70 @@
+package org.hypertrace.core.documentstore.mongo.query.parser;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.hypertrace.core.documentstore.expression.impl.AliasedIdentifierExpression;
+import org.hypertrace.core.documentstore.expression.impl.ArrayRelationalFilterExpression;
+import org.hypertrace.core.documentstore.expression.impl.DocumentArrayFilterExpression;
+import org.hypertrace.core.documentstore.expression.impl.KeyExpression;
+import org.hypertrace.core.documentstore.expression.impl.LogicalExpression;
+import org.hypertrace.core.documentstore.expression.impl.RelationalExpression;
+import org.hypertrace.core.documentstore.expression.type.FilterTypeExpression;
+import org.hypertrace.core.documentstore.mongo.MongoUtils;
+import org.hypertrace.core.documentstore.parser.FilterTypeExpressionVisitor;
+
+// Visitor for building the let clause in lookup stage from filter expressions
+class MongoLetClauseBuilder implements FilterTypeExpressionVisitor {
+
+  private final String subQueryAlias;
+
+  MongoLetClauseBuilder(String subQueryAlias) {
+    this.subQueryAlias = subQueryAlias;
+  }
+
+  @Override
+  public Map<String, Object> visit(LogicalExpression expression) {
+    Map<String, Object> letClause = new HashMap<>();
+    for (FilterTypeExpression operand : expression.getOperands()) {
+      letClause.putAll(operand.accept(this));
+    }
+    return Collections.unmodifiableMap(letClause);
+  }
+
+  @Override
+  public Map<String, Object> visit(RelationalExpression expression) {
+    Map<String, Object> letClause = new HashMap<>();
+    letClause.putAll(expression.getLhs().accept(new MongoLetClauseSelectTypeExpressionVisitor()));
+    letClause.putAll(expression.getRhs().accept(new MongoLetClauseSelectTypeExpressionVisitor()));
+    return Collections.unmodifiableMap(letClause);
+  }
+
+  private class MongoLetClauseSelectTypeExpressionVisitor extends MongoEmptySelectionTypeParser {
+
+    @Override
+    public Map<String, Object> visit(AliasedIdentifierExpression aliasedExpression) {
+      Map<String, Object> letClause = new HashMap<>();
+      if (aliasedExpression.getContextAlias().equals(subQueryAlias)) {
+        letClause.put(
+            MongoUtils.encodeVariableName(aliasedExpression.getName()),
+            MongoUtils.PREFIX + aliasedExpression.getName());
+      }
+      return Collections.unmodifiableMap(letClause);
+    }
+  }
+
+  @Override
+  public Map<String, Object> visit(KeyExpression expression) {
+    return Collections.emptyMap();
+  }
+
+  @Override
+  public Map<String, Object> visit(ArrayRelationalFilterExpression expression) {
+    return Collections.emptyMap();
+  }
+
+  @Override
+  public Map<String, Object> visit(DocumentArrayFilterExpression expression) {
```

Comment:
We don't need to override these, right?

Reply:
I just realized that it might be confusing because I put the inner class in between the outer class methods. 
Did you think that these are methods of the inner class? If so, please note that these are outer class methods and since it implements filter type expression it has to implement these methods as well.

To avoid confusion, I've moved the inner class below the outer class methods

Reply:
Oh, yeah. That was the confusion. Makes sense to keep the inner classes towards the end.